pipeline/authn/authenticator_cookie_session.go

```diff
@@ -0,0 +1,97 @@
+package authn
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/pkg/errors"
+
+	"github.com/ory/oathkeeper/driver/configuration"
+	"github.com/ory/oathkeeper/helper"
+	"github.com/ory/oathkeeper/pipeline"
+)
+
+type AuthenticatorCookieSessionConfiguration struct {
+	CookieName         string `json:"session_cookie_name"`
+	SessionStoreOrigin string `json:"session_store_origin"`
+}
+
+type AuthenticatorCookieSession struct {
+	c configuration.Provider
+}
+
+func NewAuthenticatorCookieSession(c configuration.Provider) *AuthenticatorCookieSession {
+	return &AuthenticatorCookieSession{
+		c: c,
+	}
+}
+
+func (a *AuthenticatorCookieSession) GetID() string {
+	return "cookie_session"
+}
+
+func (a *AuthenticatorCookieSession) Validate() error {
+	if !a.c.AuthenticatorCookieSessionIsEnabled() {
+		return errors.WithStack(ErrAuthenticatorNotEnabled.WithReasonf(`Authenticator "%s" is disabled per configuration.`, a.GetID()))
+	}
+
+	return nil
+}
+
+func (a *AuthenticatorCookieSession) Authenticate(r *http.Request, config json.RawMessage, _ pipeline.Rule) (*AuthenticationSession, error) {
+	var cf AuthenticatorCookieSessionConfiguration
+
+	if len(config) == 0 {
+		config = []byte("{}")
+	}
+
+	d := json.NewDecoder(bytes.NewBuffer(config))
+	d.DisallowUnknownFields()
+	if err := d.Decode(&cf); err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	sessionCookie, err := r.Cookie(cf.CookieName)
+	if err != nil {
+		return nil, errors.WithStack(ErrAuthenticatorNotResponsible)
+	}
+
+	subject, extra, err := getSession(cf, sessionCookie.Value)
+	if err != nil {
+		return nil, helper.ErrForbidden.WithReason(err.Error()).WithTrace(err)
+	}
+
+	return &AuthenticationSession{
+		Subject: subject,
+		Extra:   extra,
+	}, nil
+}
+
+func getSession(cf AuthenticatorCookieSessionConfiguration, sessionId string) (string, map[string]interface{}, error) {
+	resp, err := http.Get(cf.SessionStoreOrigin + "/sessions/" + sessionId)
```

Comment:
This makes some assumptions about the upstream store. While I think that we can't avoid it (e.g. in session) I think it should be avoided here. What do you think about a request that simply forwards the incoming request + headers to the upstream? Check out how ambassador is doing it: https://www.getambassador.io/reference/services/auth-service/#the-external-authentication-service

That would also remove the need for `cf.CookieName`

Reply:
I like that a lot better. Do you think we should send the full request through, or normalize it into a GET or HEAD request? (I'm thinking if there's a post request with a large request body like a file upload the session store probably doesn't need all that data sent to it).

Reply:
Ah I see ambassador just forwards the method + headers and has a 0 length body. That sounds good to me. Will update shortly

Reply:
One issue with removing the need for `cf.CookieName` is if the user wants a `cookie-session` authn and an `anonymous` authn we can't return `ErrAuthenticatorNotResponsible` if we don't know the cookie name to switch on. Would you prefer this happen through return status code of the session validator service or by making `CookieName` an optional option that allows fallthrough if set? I have my concerns about return status code because really the most appropriate return status code would probably be `400`, but I think it's important that any `4xx-5xx` value means invalid credentials.

Reply:
> Ah I see ambassador just forwards the method + headers and has a 0 length body. That sounds good to me. Will update shortly

Exactly :)

> One issue with removing the need for cf.CookieName is if the user wants a cookie-session authn and an anonymous authn we can't return ErrAuthenticatorNotResponsible if we don't know the cookie name to switch on. 

How about an optional filter: `only: foo,bar,baz` -  or alternatively `without: foo,bar`